src/lib.rs

```diff
@@ -72,6 +72,35 @@ impl io::Read for IoBuffer {
     }
 }
 
+impl IoBuffer {
+    /// Read a full line (terminated by `\n`), if any.  Any partial line is
+    /// left unread. The terminator is not included in the result.
+    pub fn read_full_line(&mut self) -> Option<Vec<u8>> {
+        let mut lock = self.inner.lock().expect("lock poisoned");
+        let mut p = lock.buf[lock.pos..].split(|c| *c == b'\n');
+        match p.next() {
+            Some(line) => {
+                match p.next() {
+                    Some(_rest) => {
+                        let line = line.to_vec();
+                        lock.pos += line.len() + 1;
+                        Some(line)
+                    }
+                    None => None, // incomplete line
+                }
+            }
+            None => None, // no data
```

Comment:
I feel like some `map` and `flatten` might be possible here.  `None => None` always seems unnecessary.  But perhaps that's uglier.

Reply:
I think it's clearer this way - the two `None` cases are quite distinct (no line, vs partial line). Also a `map` runs into lifetime issues.